knowledge-base/adr/0036-batch-references.md

```diff
@@ -0,0 +1,158 @@
+
+```ts
+/* request 1 */
+const createRequest = businessPartnerApi
+   .requestBuilder()
+   .create(businessPartner);
+
+const newBusinessPartner = await createRequest.execute(...);
+
+/* request 2 - create as child of with reference to newly created business partner */
+const asChildOfRequest = businessPartnerAddressApi
+.requestBuilder()
+.create(address)
+.asChildOf(newBusinessPartner, businessPartnerApi.schema.TO_BUSINESS_PARTNER_ADDRESS);
+
+const newChildAddress = await asChildOfRequest.execute(...);
+
+/* request 3 - bound function with reference to newly created business partner => let's assume there is a bound function in businessPartner */
+const functionRequest = newBusinessPartner.getNumOfAddresses();
+
+const numOfAddresses = await functionRequest.execute(...);
+```
+
+### Proposal
+
+Requests that can return entities, should allow creating objects, that act as references to an entity.
+Those requests are:
+
+- get by key
+- create
+- update
+- bound/unbound functions/actions
+
+Delete and get all probably does not make sense for this.
+
+API (`getBatchReference()`):
+
+```ts
+/* extend request type with getBatchReference() method */
+interface CreateRequestBuilder<EntityT> {
+  ...
+  getBatchReference(): BatchReference
+}
+
+interface BatchReference {
+  id: string;
+}
+```
+
+> Note that users should also be able to set the batch reference (see [Repeatability](#repeatability)).
+> This reference can be used as a parameter in other requests instead an entity. This includes:
+
+- createAsChildOf
+- delete
+- update
+- unbound functions/actions with entity parameters
+
+However, this will not work for bound functions, as they rely on the physical existence of an entity.
+
+To solve this we should have a new method on the entity builder, that allows building an entity based on a reference only.
+
+API (`fromBatchReference()`):
+
+```ts
+interface EntityBuilder<EntityT> {
+   ...
+   fromBatchRef(ref: BatchReference): EntityT
+}
+```
+
+#### Example
+
+This is the same example as described in the current state, only with batch.
+
+```ts
+/* request 1 */
+const createRequest = businessPartnerApi
+  .requestBuilder()
+  .create(businessPartner);
+
+/* reference to newly created business partner */
+const businessPartnerRef = createRequest.getBatchReference();
+
+/* request 2 - create as child of with reference to newly created business partner */
+const asChildOfRequest = businessPartnerAddressApi
+  .requestBuilder()
+  .create(address)
+  .asChildOf(
+    businessPartnerRef,
+    businessPartnerApi.schema.TO_BUSINESS_PARTNER_ADDRESS
+  );
+
+/* request 3 - bound function with reference to newly created business partner => let's assume there is a bound function in businessPartner */
+const functionRequest = businessPartnerApi.entityBuilder
+  .fromBatchReference(businessPartnerRef)
+  .doSomething();
+
+const batchResponse = await changeset(createRequest, asChildOfRequest, functionRequest).execute(...);
```

Comment:
Should this be : await batch(changeset(createRequest, asChildOfRequest), functionRequest).execute(...); ?

Reply:
Yes, almost. Will fix it 😊